File: ffi/ts/src/address.ts
import {
  Bech32,
  Bech32M,
  castToDpk,
  decodeAddress,
  encodeAddress,
  freeObj,
  getValueAsCStr,
} from './blsct'

import { DoublePublicKey } from './keys/doublePublicKey'

/**
 * Represents the available address encoding formats.
 */
export enum AddressEncoding {
  /** Bech32 encoding as defined in BIP-173. */
  Bech32,
  /** Bech32m encoding as defined in BIP-350. */
  Bech32M,
}

/**
 * Provides static utility methods for encoding and decoding addresses.
 *
 * This class is intended to be used as a static container and should not be instantiated.
 *
 * Examples:
 * ```ts
 * const { Address, DoublePublicKey, AddressEncoding } = require('navio-blsct')
 * const dpk = DoublePublicKey.random()
 * const addr = Address.encode(dpk, AddressEncoding.Bech32M)
 * addr
 * const decDpk = Address.decode(addr)
 * decDpk.serialize() === dpk.serialize()  // true
 * ```
 */
export class Address {
  /** Encode a `DoublePublicKey` to an address string using the specified encoding 
   *
   * @param addrDpk - The `DoublePublicKey` to encode.
   * @param encoding - The desired address encoding format (`Bech32` or `Bech32M`).
   * @return The encoded address as a string.
   */
  static encode(
    addrDpk: DoublePublicKey,
    encoding: AddressEncoding,
  ): string {
    let blsctEncoding: AddressEncoding = 
      encoding === AddressEncoding.Bech32 ? Bech32 : Bech32M

    const dpk = castToDpk(addrDpk.value())
    const rv = encodeAddress(dpk, blsctEncoding)
    if (rv.result !== 0) {
      const msg = `Failed to encode address. Error code = ${rv.result}`
      freeObj(rv)
      throw new Error(msg)
    }
    const addrCStr = getValueAsCStr(rv);
    freeObj(rv)

    return addrCStr;
  }

  /** Decode an address string to a `DoublePublicKey`.
   *
   * @param addrStr - The address string to decode.
   * @return A `DoublePublicKey` instance representing the decoded address.
   * @throws Error if the decoding fails.
   */
  static decode(addrStr: string): DoublePublicKey {
    const rv = decodeAddress(addrStr)
    if (rv.result !== 0) {
      const msg = `Failed to decode address. Error code = ${rv.result}`
      freeObj(rv)
      throw new Error(msg)
    }
    const addrDpk = DoublePublicKey.fromObj(rv.value)
    freeObj(rv)

    return addrDpk
  }
}

File: ffi/rust/src/address.rs
use crate::keys::double_public_key::DoublePublicKey;
use crate::{
  blsct_obj::{BlsctObj, self},
  ffi::{
    AddressEncoding,
    decode_address,
    encode_address,
    free_obj,
  },
};
use std::ffi::{
  c_char,
  c_void,
  CStr,
  CString,
  NulError,
};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum Error<'a> {
  #[error("Failed to allocate memory for {0:?}")]
  FailedToAllocateMemory(&'static str),
  
  #[error("Failed to construct BlsctRetVal: {0:?}")]
  FailedToConstructBlsctRetVal(blsct_obj::Error<'a>),

  #[error("Failed to encode address: {0:?}")]
  FailedToEncodeAddress(&'a DoublePublicKey),

  #[error("Failed to decode address: {0:?}")]
  FailedToDecodeAddress(&'a str),

  #[error("Failed to construct CString: {0:?}")]
  FailedToConstructCString(NulError),
}

pub struct Address();

impl Address {
  pub fn encode(
    addr_dpk: &DoublePublicKey,
    encoding: AddressEncoding,
  ) -> Result<String, Error> {
    let rv = unsafe { encode_address(addr_dpk.value(), encoding) };
    if rv.is_null() {
      unsafe { free_obj(rv as *mut c_void) };
      Err(Error::FailedToAllocateMemory("BlsctRetVal"))

    } else if unsafe { (*rv).result != 0 } {
      unsafe { free_obj(rv as *mut c_void) };
      Err(Error::FailedToEncodeAddress(addr_dpk))

    } else {
      let addr_c_str = unsafe { CStr::from_ptr((*rv).value as *const c_char) };
      let addr = addr_c_str.to_str().unwrap().to_string();
      unsafe { free_obj((*rv).value as *mut c_void) };
      Ok(addr)
    }
  }

  pub fn decode(addr_str: &str) -> Result<DoublePublicKey, Error> {
    let c_addr_str = CString::new(addr_str)
      .map_err(|e| Error::FailedToConstructCString(e))?;

    let rv = unsafe { decode_address(c_addr_str.as_ptr()) };
    if rv.is_null() {
      unsafe { free_obj(rv as *mut c_void) };
      Err(Error::FailedToAllocateMemory("BlsctRetVal"))

    } else if unsafe { (*rv).result != 0 } {
      unsafe { free_obj(rv as *mut c_void) };
      Err(Error::FailedToDecodeAddress(addr_str))

    } else {
      let addr_dpk = BlsctObj::from_retval(rv)
        .map_err(|e| Error::FailedToConstructBlsctRetVal(e))?;
      Ok(addr_dpk.into())
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::{
    initializer::init,
    keys::double_public_key::DoublePublicKey,
  };

  #[test]
  fn test_encode_decode() {
    init();
    for encoding in [AddressEncoding::Bech32, AddressEncoding::Bech32M] {
      let addr_dpk = DoublePublicKey::random().unwrap();
      let addr_str = Address::encode(&addr_dpk, encoding).unwrap();
      println!("addr_str: {}", addr_str);

      let decoded_dpk = Address::decode(&addr_str).unwrap();
      assert_eq!(addr_dpk, decoded_dpk);
    }
  }
}

